### PR TITLE
Introduced `archngv` package and its dependencies

### PR DESCRIPTION
**Pull request by arsenius7**
* py-jenkspy
 * py-morphmath
 * py-morphspatial
 * py-numpy-stl
 * py-openmesh
 * py-python-utils
 * py-spatial-index
 * py-tess
 * py-trimesh

updated:

 * py-munkres -> 1.1.2
 * py-tmd -> 2.0.4
 * py-tns -> `@space2`

### CODE REVIEW

**Comment by arsenius7** [comment]
The error message states that:
> As of version 1.1.0, this munkres package no longer supports Python 2. Either upgrade to Python 3.5 or better, or use an older version of munkres (e.g., 1.0.12).

What is our general policy here, in case we don't care about Python 2 compatibility for a given package (i.e. it is expected to be used only from Python 3 environments)?

Do we still try to distinguish the version requirements for Python 2 and 3, or I can explicitly specify that a given package is not available for Python 2 (and thus should not be tested)?

**Comment by matz-e** [comment]
> The error message states that:
> 
> > As of version 1.1.0, this munkres package no longer supports Python 2. Either upgrade to Python 3.5 or better, or use an older version of munkres (e.g., 1.0.12).
> 
> What is our general policy here, in case we don't care about Python 2 compatibility for a given package (i.e. it is expected to be used only from Python 3 environments)?
> 
> Do we still try to distinguish the version requirements for Python 2 and 3, or I can explicitly specify that a given package is not available for Python 2 (and thus should not be tested)?

I think we could make the `depends_on("python@3.5:", when="@:1.…` dance here? I.e. have an older version for python2.

In general, I think that we should ditch Python 2 support for all applications that expose no internals to users. Spykfunc has moved to Python 3 only already, and I have no intention of adding support for Python 2 back.

**Comment by arsenius7** [comment]
> I think we could make the depends_on("python@3.5:", when="@:1.… dance here? I.e. have an older version for python2.

That would make CI pass, but will end up in a runtime error if someone tries to use `archngv^python@2.7` (not like it matters much, so I'll go this way).

> In general, I think that we should ditch Python 2 support for all applications that expose no internals to users.

+1.
I just wanted to know how it is done technically (just ignoring Python 2 versions being broken, or explicitly setting some "Python 3 only" flags).

**Comment by matz-e** [comment]
Spykfunc for example has an explicit dependency on Python 3:
```
    depends_on('python@3.6:')
```

**Comment by arsenius7** [comment]
@matz-e , @pramodk , I can not reproduce the latest CI error locally; and thus would appreciate some help with deciphering it.
Thanks.

**Comment by pramodk** [comment]
most likely due to blas/lapack dependencies, will take a look with @matz-e

**Comment by pramodk** [comment]
@arsenius7 : I couldn't push to your branch, could you try : 

```
→ git show HEAD
commit fbbdcb55c81c3be58b3b1e73c81b2f1e568ae04d (HEAD -> archngv, origin/arseny/archngv)
Author: Pramod S Kumbhar <pramod.s.kumbhar@gmail.com>
Date:   Fri May 31 08:21:59 2019 +0200

    try building with blas/lapack

diff --git a/deploy/packages/bbp-packages.yaml b/deploy/packages/bbp-packages.yaml
index cf7e35c6b..4111d6379 100644
--- a/deploy/packages/bbp-packages.yaml
+++ b/deploy/packages/bbp-packages.yaml
@@ -21,7 +21,6 @@ packages:
       - compiler
       - python
     specs:
-      - archngv@0.0.0
       - brainbuilder@0.11.1
       - connectome-tools@0.3.2
       - spykfunc@0.13.2
@@ -54,6 +53,17 @@ packages:
     specs:
       - placement-algorithm@2.0.8

+  gnu-stable-lapack-python-serial:
+    target_matrix:
+      - gnu-stable
+    requires:
+      - architecture
+      - compiler
+      - lapack
+      - python
+    specs:
+      - archngv@0.0.0
+
```

**Comment by arsenius7** [comment]
> Error: Conflicting conditional dependencies onpackage py-tmd for spec py-tmd@2.0.3: ^blas ^bzip2 ^diffutils ^gdbm ^lapack ^ncurses ^openssl ^perl@5.14.0: ^pkgconfig ^py-enum34@1.0.4: ^py-numpy@1.13:+blas+lapack ^py-scipy@1.0: ^py-setuptools@0.6c5: ^python@2.7:2.8,3.4: ^readline ^sqlite ^zlib@1.1.2:

The good news is that I'm getting the same thing locally now :), although I still have no clue how to deal with it.
Is it 
```
depends_on('py-munkres@1.0.11', type='run', when='^python@:3.4.99')
depends_on('py-munkres@1.0.12:', type='run', when='^python@3.5:')
```
that is problematic?

**Comment by matz-e** [comment]
> > Error: Conflicting conditional dependencies onpackage py-tmd for spec py-tmd@2.0.3: ^blas ^bzip2 ^diffutils ^gdbm ^lapack ^ncurses ^openssl ^perl@5.14.0: ^pkgconfig ^py-enum34@1.0.4: ^py-numpy@1.13:+blas+lapack ^py-scipy@1.0: ^py-setuptools@0.6c5: ^python@2.7:2.8,3.4: ^readline ^sqlite ^zlib@1.1.2:
> 
> The good news is that I'm getting the same thing locally now :), although I still have no clue how to deal with it.
> Is it
> 
> ```
> depends_on('py-munkres@1.0.11', type='run', when='^python@:3.4.99')
> depends_on('py-munkres@1.0.12:', type='run', when='^python@3.5:')
> ```
> 
> that is problematic?

I would take that out first and see if it resolves things.

**Comment by arsenius7** [comment]
OK, at the cost of dropping Python<3.5 support for `py-tmd`, the CI is passing now.
I have rolled back `gnu-stable-lapack-python-serial` change, as it does not seem to affect the result (at least for the deployment?).
The resulting module from PR is working as well.

**Review by matz-e — APPROVED** [review]
Looks good to me now. Maybe @pramodk has more comments?

**Comment by pramodk** [comment]
Using upstream develop I get error message like below. I am not sure why it try to merge those python versions :

```
 → spack spec py-tmd
Input spec
--------------------------------
py-tmd

Concretized
--------------------------------
==> Error: Conflicting conditional dependencies for spec

	py-tmd ^hdf5@1.8.4:+hl ^pkgconfig ^py-cython@0.23: ^py-enum34@1.0: ^py-h5py@2.8: ^py-nose@1.0: ^py-numpy@1.7: ^py-pkgconfig ^py-setuptools ^py-six ^python@2.7:2.8,3.4: ^zlib@1.1.2:

Cannot merge constraint

	py-munkres@1.0.12:

into

	py-munkres@1.0.11
```

> this error message should come from spack/spack/pull/9975

Seems like if we help concretisation by explicitly specifying python version, it works. This should happen in Jenkins CI and hence I was not expecting an error :

```
→ git diff
diff --git a/var/spack/repos/builtin/packages/py-tmd/package.py b/var/spack/repos/builtin/packages/py-tmd/package.py
index ee4fb2508..4021a5fce 100644
--- a/var/spack/repos/builtin/packages/py-tmd/package.py
+++ b/var/spack/repos/builtin/packages/py-tmd/package.py
@@ -19,6 +19,7 @@ class PyTmd(PythonPackage):

     depends_on('py-enum34@1.0:', type='run', when='^python@:3.3.99')
     depends_on('py-h5py@2.8:', type='run')
+    depends_on('py-munkres@1.0.11', type='run', when='^python@:3.4.99')
     depends_on('py-munkres@1.0.12:', type='run', when='^python@3.5:')
     depends_on('py-numpy@1.8:', type='run')
     depends_on('py-scikit-learn@0.19:', type='run')

± |archngvv {20} U:1 ?:29 ✗| → spack spec py-tmd
Input spec
--------------------------------
py-tmd

Concretized
--------------------------------
==> Error: Conflicting conditional dependencies onpackage py-tmd for spec py-tmd ^bzip2 ^diffutils ^gdbm ^ncurses ^openssl ^perl@5.14.0: ^pkgconfig ^py-cython@0.23: ^py-enum34@1.0: ^py-h5py@2.8: ^py-nose@1.0: ^py-numpy@1.7: ^py-pkgconfig ^py-setuptools ^py-six ^python@2.7:2.8,3.4: ^readline ^sqlite ^zlib

 2019-06-04 07:13:24 ☆  bluebrain355 in ~/workarena/software/sources/spack
± |archngvv {20} U:1 ?:29 ✗| → spack spec py-tmd ^python@3.6.0
Input spec
--------------------------------
py-tmd
    ^python@3.6.0

Concretized
--------------------------------
py-tmd@2.0.4%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-h5py@2.9.0%clang@9.0.0-apple+mpi arch=darwin-mojave-x86_64
        ^hdf5@1.8.17%clang@9.0.0-apple~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=darwin-mojave-x86_64
        ^mpich@3.2%clang@9.0.0-apple device=ch3 +hydra netmod=tcp patches=e591891fae6a1d029b95edbc93865c57d165784e77cf8b5822ae7d014b137a03 +pmi+romio~verbs arch=darwin-mojave-x86_64
        ^py-cython@0.29%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^python@3.6.0%clang@9.0.0-apple+dbm~optimizations patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+pythoncmd~shared~tk~ucs4 arch=darwin-mojave-x86_64
                ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-mojave-x86_64
                    ^diffutils@3.6%clang@9.0.0-apple arch=darwin-mojave-x86_64
                ^gdbm@1.18.1%clang@9.0.0-apple arch=darwin-mojave-x86_64
                    ^readline@7.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
                        ^ncurses@6.1%clang@9.0.0-apple~symlinks~termlib arch=darwin-mojave-x86_64
                            ^pkg-config@develop%clang@9.0.0-apple+internal_glib arch=darwin-mojave-x86_64
                ^openssl@develop%clang@9.0.0-apple+systemcerts arch=darwin-mojave-x86_64
                ^sqlite@3.25.3%clang@9.0.0-apple~functions arch=darwin-mojave-x86_64
                ^zlib@1.2.8%clang@9.0.0-apple+optimize+pic+shared arch=darwin-mojave-x86_64
        ^py-mpi4py@3.0.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^py-setuptools@40.4.3%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-numpy@1.15.2%clang@9.0.0-apple+blas+lapack arch=darwin-mojave-x86_64
            ^openblas@0.3.3%clang@9.0.0-apple cpu_target= ~ilp64 patches=714aea33692304a50bd0ccde42590c176c82ded4a8ac7f06e573dc8071929c33,f1b066a4481a50678caeb7656bf3e6764f45619686ac465f257c8017a2dc1ff0 +pic+shared threads=none ~virtual_machine arch=darwin-mojave-x86_64
        ^py-pkgconfig@1.2.2%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^py-nose@1.3.7%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-six@1.11.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-munkres@1.1.2%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-scikit-learn@0.20.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-scipy@1.1.0%clang@9.0.0-apple arch=darwin-mojave-x86_64


± |archngvv {20} U:1 ?:29 ✗| → spack spec py-tmd ^python@3.4.0
Input spec
--------------------------------
py-tmd
    ^python@3.4.0

Concretized
--------------------------------
py-tmd@2.0.4%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-h5py@2.9.0%clang@9.0.0-apple+mpi arch=darwin-mojave-x86_64
        ^hdf5@1.8.17%clang@9.0.0-apple~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=darwin-mojave-x86_64
        ^mpich@3.2%clang@9.0.0-apple device=ch3 +hydra netmod=tcp patches=e591891fae6a1d029b95edbc93865c57d165784e77cf8b5822ae7d014b137a03 +pmi+romio~verbs arch=darwin-mojave-x86_64
        ^py-cython@0.29%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^python@3.4.0%clang@9.0.0-apple+dbm~optimizations patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+pythoncmd~shared~tk~ucs4 arch=darwin-mojave-x86_64
                ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-mojave-x86_64
                    ^diffutils@3.6%clang@9.0.0-apple arch=darwin-mojave-x86_64
                ^gdbm@1.18.1%clang@9.0.0-apple arch=darwin-mojave-x86_64
                    ^readline@7.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
                        ^ncurses@6.1%clang@9.0.0-apple~symlinks~termlib arch=darwin-mojave-x86_64
                            ^pkg-config@develop%clang@9.0.0-apple+internal_glib arch=darwin-mojave-x86_64
                ^openssl@develop%clang@9.0.0-apple+systemcerts arch=darwin-mojave-x86_64
                ^sqlite@3.25.3%clang@9.0.0-apple~functions arch=darwin-mojave-x86_64
                ^zlib@1.2.8%clang@9.0.0-apple+optimize+pic+shared arch=darwin-mojave-x86_64
        ^py-mpi4py@3.0.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^py-setuptools@40.4.3%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-numpy@1.15.2%clang@9.0.0-apple+blas+lapack arch=darwin-mojave-x86_64
            ^openblas@0.3.3%clang@9.0.0-apple cpu_target= ~ilp64 patches=714aea33692304a50bd0ccde42590c176c82ded4a8ac7f06e573dc8071929c33,f1b066a4481a50678caeb7656bf3e6764f45619686ac465f257c8017a2dc1ff0 +pic+shared threads=none ~virtual_machine arch=darwin-mojave-x86_64
        ^py-pkgconfig@1.2.2%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^py-nose@1.3.7%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-six@1.11.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-munkres@1.0.11%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-scikit-learn@0.20.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-scipy@1.1.0%clang@9.0.0-apple arch=darwin-mojave-x86_64

± |archngvv {20} U:1 ?:29 ✗| → spack spec py-tmd ^python@2.7.0
Input spec
--------------------------------
py-tmd
    ^python@2.7.0

Concretized
--------------------------------
py-tmd@2.0.4%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-enum34@1.1.6%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-setuptools@40.4.3%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^python@2.7.0%clang@9.0.0-apple+dbm~optimizations patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+pythoncmd~shared~tk~ucs4 arch=darwin-mojave-x86_64
                ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-mojave-x86_64
                    ^diffutils@3.6%clang@9.0.0-apple arch=darwin-mojave-x86_64
                ^gdbm@1.18.1%clang@9.0.0-apple arch=darwin-mojave-x86_64
                    ^readline@7.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
                        ^ncurses@6.1%clang@9.0.0-apple~symlinks~termlib arch=darwin-mojave-x86_64
                            ^pkg-config@develop%clang@9.0.0-apple+internal_glib arch=darwin-mojave-x86_64
                ^openssl@develop%clang@9.0.0-apple+systemcerts arch=darwin-mojave-x86_64
                ^sqlite@3.25.3%clang@9.0.0-apple~functions arch=darwin-mojave-x86_64
                ^zlib@1.2.8%clang@9.0.0-apple+optimize+pic+shared arch=darwin-mojave-x86_64
    ^py-h5py@2.9.0%clang@9.0.0-apple+mpi arch=darwin-mojave-x86_64
        ^hdf5@1.8.17%clang@9.0.0-apple~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=darwin-mojave-x86_64
        ^mpich@3.2%clang@9.0.0-apple device=ch3 +hydra netmod=tcp patches=e591891fae6a1d029b95edbc93865c57d165784e77cf8b5822ae7d014b137a03 +pmi+romio~verbs arch=darwin-mojave-x86_64
        ^py-cython@0.29%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-mpi4py@3.0.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-numpy@1.15.2%clang@9.0.0-apple+blas+lapack arch=darwin-mojave-x86_64
            ^openblas@0.3.3%clang@9.0.0-apple cpu_target= ~ilp64 patches=714aea33692304a50bd0ccde42590c176c82ded4a8ac7f06e573dc8071929c33,f1b066a4481a50678caeb7656bf3e6764f45619686ac465f257c8017a2dc1ff0 +pic+shared threads=none ~virtual_machine arch=darwin-mojave-x86_64
        ^py-pkgconfig@1.2.2%clang@9.0.0-apple arch=darwin-mojave-x86_64
            ^py-nose@1.3.7%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-six@1.11.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-munkres@1.0.11%clang@9.0.0-apple arch=darwin-mojave-x86_64
    ^py-scikit-learn@0.20.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
        ^py-scipy@1.1.0%clang@9.0.0-apple arch=darwin-mojave-x86_64
```

By removing `depends_on('py-munkres@1.0.11', type='run', when='^python@:3.4.99')`, we are removing python2 support?.  Ok for me unless users of this package start asking Python2 support.

> Note that on HBP machines we are still deploying python2. But it's not related here.

**Comment by arsenius7** [comment]
> By removing depends_on('py-munkres@1.0.11', type='run', when='^python@:3.4.99'), we are removing python2 support?

More like breaking it, in fact, because on using `py-tmd` one would get a message about unsatisfied `pip` dependency. :(

Can I instead explicitly specify that `py-munkres` is Python 3.6+ only, by adding
```
depends_on('python@3.6:')
```
to `py-munkres` recipe; or it's not a practice to be used with _library_ packages?

**Comment by matz-e** [comment]
> > By removing depends_on('py-munkres@1.0.11', type='run', when='^python@:3.4.99'), we are removing python2 support?
> 
> More like breaking it, in fact, because on using `py-tmd` one would get a message about unsatisfied `pip` dependency. :(
> 
> Can I instead explicitly specify that `py-munkres` is Python 3.6+ only, by adding
> 
> ```
> depends_on('python@3.6:')
> ```
> 
> to `py-munkres` recipe; or it's not a practice to be used with _library_ packages?

I think that's valid practice for libraries, too. In fact, you could make that version dependent, if that is the underlying issue here (1.0.11 for py2, otherwise py3).

**Comment by arsenius7** [comment]
> In fact, you could make that version dependent, if that is the underlying issue here (1.0.11 for py2, otherwise py3).

Good to know; but as these conditional dependencies will in turn leak to packages dependent on `py-munkres`, I guess it's not worth it this time.

I'll limit `py-munkres` to Python 3.6+ for the moment, if it does not break other deployment locations, so that if / when someone tries to use it with Python 2.7, the error message is explicit.

**Comment by arsenius7** [comment]
retest please?

**Comment by matz-e** [comment]
Retest this please

**Comment by arsenius7** [comment]
OK, obviously CI is not happy with limiting Python version for a library. ^^

**Comment by matz-e** [comment]
> OK, obviously CI is not happy with limiting Python version for a library. ^^

Well, we seemingly explicitly request some software with py2 that needs a py3 library. Refusal to do such things seems to be the right way.

**Comment by arsenius7** [comment]
No doubt ^^; the question is: should I remove the thing that does this explicit request (how can I find it? the only final application I can think of is `placement-algorithm`); or get back to "broken `py-tmd` package passing CI".

**Comment by arsenius7** [comment]
Is `deploy/packages/python-packages.yaml` the place to look at?

**Comment by matz-e** [comment]
> Is `deploy/packages/python-packages.yaml` the place to look at?

Yes. Let me see if I can push to this PR

**Comment by matz-e** [comment]
Meh. Try this:
```
--- a/deploy/packages/python-packages.yaml
+++ b/deploy/packages/python-packages.yaml
@@ -21,6 +21,7 @@ packages:
       - python
     blacklist:
       - 'py-bb5 ^python@2.7.15'
+      - 'py-munkres ^python@2.7.15'
     specs:
       - py-argcomplete
       - py-argparse
@@ -76,7 +77,6 @@ packages:
       # - py-neo
       # - py-networkx
       # - py-nose
-      - py-munkres
       - py-neurom
       - py-pandas
       - py-perf
```

File: deploy/packages/python-packages.yaml [hunk 1]
```diff
@@ -76,7 +77,6 @@ packages:
       # - py-neo
       # - py-networkx
       # - py-nose
-      - py-munkres
```
Comment (matz-e) [inline deploy/packages/python-packages.yaml]: Ah sorry! Shouldn't have removed this line. `munkres` can stay in here, and the blacklist should make sure that it's not build for py2.

Comment (arsenius7) [reply deploy/packages/python-packages.yaml]: Serves me right for blindly copying the diff, without listening to own doubts. :)

**Pull request merged**